src/files-and-uploads/FilesAndUploads.jsx

```diff
@@ -0,0 +1,273 @@
+
+import { RequestStatus } from '../data/constants';
+import { useModels } from '../generic/model-store';
+import {
+  addAssetFile,
+  deleteAssetFile,
+  fetchAssets,
+  updateAssetLock,
+} from './data/thunks';
+import messages from './messages';
+
+import FileInput, { fileInput } from './FileInput';
+import FilesAndUploadsProvider from './FilesAndUploadsProvider';
+import {
+  GalleryCard,
+  ListCard,
+  TableActions,
+} from './table-components';
+import ApiStatusToast from './ApiStatusToast';
+
+const FilesAndUploads = ({
+  courseId,
+  // injected
+  intl,
+}) => {
+  const dispatch = useDispatch();
+  const defaultVal = 'card';
+  const columnSizes = { xs: 12, sm: 6, lg: 3 };
+  const [currentView, setCurrentView] = useState(defaultVal);
+  const [isDeleteOpen, setDeleteOpen, setDeleteClose] = useToggle(false);
+  const [isAddOpen, setAddOpen, setAddClose] = useToggle(false);
+  const [selectedRowCount, setSelectedRowCount] = useState(0);
+  const assetIds = useSelector(state => state.assets.assetIds);
+
+  useEffect(() => {
+    dispatch(fetchAssets(courseId));
+  }, [courseId]);
+  const totalCount = useSelector(state => state.assets.totalCount);
+  const addAssetStatus = useSelector(state => state.assets.addingStatus);
+  const deleteAssetStatus = useSelector(state => state.assets.deletingStatus);
+  const saveAssetStatus = useSelector(state => state.assets.savingStatus);
+  const loadingStatus = useSelector(state => state.assets.loadingStatus);
+  const errorMessages = useSelector(state => state.assets.errors);
+  const fileInputControl = fileInput({
+    onAddFile: (file) => dispatch(addAssetFile(courseId, file, totalCount)),
+    setSelectedRowCount,
+    setAddOpen,
+  });
+  const assets = useModels('assets', assetIds);
+
+  const handleDropzoneAsset = ({ fileData, handleError }) => {
+    try {
+      const file = fileData.get('file');
+      dispatch(addAssetFile(courseId, file, totalCount));
+    } catch (error) {
+      handleError(error);
+    }
+  };
+
+  const handleBulkDelete = (selectedFlatRows) => {
+    setSelectedRowCount(selectedFlatRows.length);
+    setDeleteOpen();
+    const assetIdsToDelete = selectedFlatRows.map(row => row.original.id);
+    assetIdsToDelete.forEach(id => dispatch(deleteAssetFile(courseId, id, totalCount)));
+  };
+
+  const handleBulkDownload = (selectedFlatRows) => {
+    selectedFlatRows.forEach(row => {
+      const { externalUrl } = row.original;
+      const link = document.createElement('a');
+      link.target = '_blank';
+      link.download = true;
+      link.href = externalUrl;
+      link.click();
+    });
+    /* ********** TODO ***********
+     * implement a zip file function when there are multiple files
+    */
+  };
+
+  const handleLockedAsset = (assetId, locked) => {
+    dispatch(updateAssetLock({ courseId, assetId, locked }));
+  };
+
+  const headerActions = ({ selectedFlatRows }) => {
+    if (_.isEmpty(assets)) {
+      return [];
```

Comment:
Why are we returning a list or a react component here? I know JS isn't type safe, but that seems risky.

Reply:
Thanks for the catch! The parameter that is calling the function expects an array of components or a functional component. But the returned empty array is no longer needed because of how I control the view of the table control bar.